controllers/replication.storage/volumereplication_controller.go

```diff
@@ -203,17 +228,32 @@ func (r *VolumeReplicationReconciler) Reconcile(ctx context.Context, req ctrl.Re
 
 			return reconcile.Result{}, err
 		}
+		switch instance.Spec.DataSource.Kind {
+		case pvcDataSource:
+			err = r.annotatePVCWithOwner(ctx, logger, req.Name, pvc)
+			if err != nil {
+				logger.Error(err, "Failed to annotate PVC owner")
+				return ctrl.Result{}, err
+			}
 
-		err = r.annotatePVCWithOwner(ctx, logger, req.Name, pvc)
-		if err != nil {
-			logger.Error(err, "Failed to annotate PVC owner")
-			return ctrl.Result{}, err
-		}
+			if err = r.addFinalizerToPVC(logger, pvc); err != nil {
+				logger.Error(err, "Failed to add PersistentVolumeClaim finalizer")
 
-		if err = r.addFinalizerToPVC(logger, pvc); err != nil {
-			logger.Error(err, "Failed to add PersistentVolumeClaim finalizer")
+				return reconcile.Result{}, err
+			}
+		case volumeGroupReplicationDataSource:
+			err = r.annotateVolumeGroupReplicationWithOwner(ctx, logger, req.Name, vgr)
+			if err != nil {
+				logger.Error(err, "Failed to annotate VolumeGroupReplication owner")
 
-			return reconcile.Result{}, err
+				return ctrl.Result{}, err
+			}
+
+			if err = r.addFinalizerToVGR(logger, vgr); err != nil {
```

Comment:
Also, why are we adding VGR finalizer here?
Shouldn't the VGR controller take care of this?

Reply:
As I mentioned above this is a safe guard check to ensure that the whole flow is followed before deleting the VR as we need to have a VGR object.